### PR TITLE
Provide Application Insights converters, telemetry type filter and configuration extension

### PR DESCRIPTION
**Pull request by tomkerkhove**
- [X] Provide Application Insights converters
- [X] Provide telemetry type filter
- [X] Provide configuration extension to wire-up the Application Insights sink
- [X] Provide docs for filter
- [X] Provide docs for sink

As discussed with @stijnmoreels, the idea is to have the basics in and add variety of tests in a follow-up PR.

Relates to #14
Relates to #15
Relates to #16
Relates to #17
Relates to #22

### CODE REVIEW

File: src/Arcus.Observability.Telemetry.Core/ContextProperties.cs [hunk 1]
```diff
@@ -3,6 +3,11 @@ namespace Arcus.Observability.Telemetry.Core
 {
     public static class ContextProperties
     {
+        public class Correlation
```
Comment (stijnmoreels) [inline src/Arcus.Observability.Telemetry.Core/ContextProperties.cs]: Static class?

File: src/Arcus.Observability.Telemetry.Serilog.Filters/TelemetryTypeFilter.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
+﻿using Arcus.Observability.Telemetry.Core;
+using Serilog.Core;
+using Serilog.Events;
+
+namespace Arcus.Observability.Telemetry.Serilog.Filters
+{
+    public class TelemetryTypeFilter : ILogEventFilter
```
Comment (stijnmoreels) [inline src/Arcus.Observability.Telemetry.Serilog.Filters/TelemetryTypeFilter.cs]: Maybe an extra XML explaination here?

File: src/Arcus.Observability.Telemetry.Serilog.Sinks.ApplicationInsights/Converters/ApplicationInsightsTelemetryConverter.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+        private readonly HttpDependencyTelemetryConverter _httpDependencyTelemetryConverter =
+            new HttpDependencyTelemetryConverter();
+
+        private readonly SqlDependencyTelemetryConverter _sqlDependencyTelemetryConverter =
+            new SqlDependencyTelemetryConverter();
+
+        public override IEnumerable<ITelemetry> Convert(LogEvent logEvent, IFormatProvider formatProvider)
+        {
+            if (logEvent.MessageTemplate.Text.StartsWith(MessagePrefixes.RequestViaHttp))
```
Comment (stijnmoreels) [inline src/Arcus.Observability.Telemetry.Serilog.Sinks.ApplicationInsights/Converters/ApplicationInsightsTelemetryConverter.cs]: Arcus code guidelines states we have braces here, no?

Comment (stijnmoreels) [reply src/Arcus.Observability.Telemetry.Serilog.Sinks.ApplicationInsights/Converters/ApplicationInsightsTelemetryConverter.cs]: I'll have to check, but switch/pattern-matching is maybe already ported from  F# to C# here.

File: src/Arcus.Observability.Telemetry.Serilog/Extensions/LogEventPropertyValueExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        /// <param name="propertyKey">Key of the property to return</param>
+        public static IReadOnlyDictionary<ScalarValue, LogEventPropertyValue> GetAsDictionary(this IReadOnlyDictionary<string, LogEventPropertyValue> eventPropertyValues, string propertyKey)
+        {
+            Guard.NotNull(eventPropertyValues, nameof(eventPropertyValues));
+
+            var logEventPropertyValue = eventPropertyValues.GetValueOrDefault(propertyKey);
+
+            if (logEventPropertyValue is DictionaryValue == false)
+            {
+                throw new NotSupportedException();
```
Comment (stijnmoreels) [inline src/Arcus.Observability.Telemetry.Serilog/Extensions/LogEventPropertyValueExtensions.cs]: Some exception explaination here would be useful.

File: src/Arcus.Observability.Telemetry.Serilog/Extensions/LogEventPropertyValueExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+            var logEventPropertyValue = eventPropertyValues.GetAsRawString(propertyKey);
+            var value = bool.Parse(logEventPropertyValue);
+            return value;
+        }
+
+        /// <summary>
+        ///     Provide a decent string representation of the event property value
+        /// </summary>
+        /// <remarks>The built-in <c>ToString</c> wraps the string with quotes</remarks>
```
Comment (stijnmoreels) [inline src/Arcus.Observability.Telemetry.Serilog/Extensions/LogEventPropertyValueExtensions.cs]: 👍 good to include this in the remarks.

File: docs/features/writing-different-telemetry-types.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+```csharp
+var telemetryContext = new Dictionary<string, object>
+{
+    {"Customer", "Arcus"},
+    {"OrderId", "ABC"},
+};
+
+logger.LogEvent("Order Created", telemetryContext);
+// Output: ""
```
Comment (stijnmoreels) [inline docs/features/writing-different-telemetry-types.md]: This output is probably not correct?

File: docs/features/telemetry-filter.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+## Installation
+
+This feature requires to install our NuGet package
+
+```shell
+PM > Install-Package Arcus.Observability.Telemetry.Serilog.Filters
+```
+
+## Telemetry Type Filter
```
Comment (stijnmoreels) [inline docs/features/telemetry-filter.md]: 1st place on clear/simple docs. 🥇

Comment (tomkerkhove) [reply docs/features/telemetry-filter.md]: Thanks man!

File: docs/features/writing-different-telemetry-types.md (hunk 6 above)
Comment (tomkerkhove) [reply docs/features/writing-different-telemetry-types.md]: Updated!

File: src/Arcus.Observability.Telemetry.Core/Extensions/IReadOnlyDictionaryExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+namespace System.Collections.Generic
+{
+    public static class IReadOnlyDictionaryExtensions
+    {
+        /// <summary>
+        ///     Provides value or default for a given dictionary entry
+        /// </summary>
+        /// <param name="dictionary">Dictionary containing data</param>
+        /// <param name="propertyKey">Key of the dictionary entry to return</param>
+        public static TValue GetValueOrDefault<TValue>(this IReadOnlyDictionary<string, TValue> dictionary, string propertyKey)
```
Comment (stijnmoreels) [inline src/Arcus.Observability.Telemetry.Core/Extensions/IReadOnlyDictionaryExtensions.cs]: Hmm, isn't `.TryGetValue` a method of the `IDictionary` type? Maybe we can make this more generilized?

File: src/Arcus.Observability.Telemetry.Serilog.Enrichers/Extensions/LogEventPropertyValueExtensions.cs [hunk 9]
```diff
@@ -55,9 +55,9 @@ public static string GetAsRawString(this IReadOnlyDictionary<string, LogEventPro
             var logEventPropertyValue = eventPropertyValues.GetValueOrDefault(propertyKey);
             var rawEnum = logEventPropertyValue?.ToDecentString();
 
-            if (Enum.TryParse(typeof(TEnum), rawEnum, out var enumRepresentation))
+            if (Enum.TryParse(rawEnum, out TEnum enumRepresentation))
```
Comment (stijnmoreels) [inline src/Arcus.Observability.Telemetry.Serilog.Enrichers/Extensions/LogEventPropertyValueExtensions.cs]: 👍

File: src/Arcus.Observability.Telemetry.Core/Extensions/IReadOnlyDictionaryExtensions.cs (hunk 8 above)
Comment (tomkerkhove) [reply src/Arcus.Observability.Telemetry.Core/Extensions/IReadOnlyDictionaryExtensions.cs]: It doesn't give you the default value if it's not there so they are not identical?

Comment (stijnmoreels) [reply src/Arcus.Observability.Telemetry.Core/Extensions/IReadOnlyDictionaryExtensions.cs]: No, I mean use the `IDictionary` type in the signature:

```csharp
GetValueOrDefault<TValue>(this IDictionary dictionary...
```

Comment (tomkerkhove) [reply src/Arcus.Observability.Telemetry.Core/Extensions/IReadOnlyDictionaryExtensions.cs]: Ok can do that as well, will change this.

Comment (tomkerkhove) [reply src/Arcus.Observability.Telemetry.Core/Extensions/IReadOnlyDictionaryExtensions.cs]: Good point!

Comment (tomkerkhove) [reply src/Arcus.Observability.Telemetry.Core/Extensions/IReadOnlyDictionaryExtensions.cs]: Actually, `IReadOnlyDictionary` and `IDictionary` are not the same so I cannot use that

Comment (stijnmoreels) [reply src/Arcus.Observability.Telemetry.Core/Extensions/IReadOnlyDictionaryExtensions.cs]: Ok, it's fine.

**Pull request merged**